File: source/ResourceLibrary/Controllers/applicationController.vb

Option Explicit On
Option Strict On

Imports Contensive.BaseClasses

Namespace Controllers
    '
    '====================================================================================================
    ''' <summary>
    ''' 
    ''' </summary>
    ''' <remarks></remarks>
    Public Class applicationController
        Implements IDisposable
        '
        ' privates passed in, do not dispose
        '
        Private cp As CPBaseClass
        '
        Public ReadOnly Property allowPlace As Boolean
        '
        Public ReadOnly Property topFolderPath As String
        '
        Public ReadOnly Property AllowGroupAdd As Boolean
        '
        '====================================================================================================
        ''' <summary>
        ''' Errors accumulated during rendering.
        ''' </summary>
        ''' <returns></returns>
        Public Property packageErrorList As New List(Of packageErrorClass)
        '
        '====================================================================================================
        ''' <summary>
        ''' data accumulated during rendering
        ''' </summary>
        ''' <returns></returns>
        Public Property packageNodeList As New List(Of packageNodeClass)
        '
        '====================================================================================================
        ''' <summary>
        ''' list of name/time used to performance analysis
        ''' </summary>
        ''' <returns></returns>
        Public Property packageProfileList As New List(Of packageProfileClass)

        '
        '====================================================================================================
        ''' <summary>
        ''' get the serialized results
        ''' </summary>
        ''' <returns></returns>
        Public Function getSerializedPackage() As String
            Dim result As String = ""
            Try
                result = serializeObject(cp, New packageClass With {
                    .success = packageErrorList.Count.Equals(0),
                    .nodeList = packageNodeList,
                    .errorList = packageErrorList,
                    .profileList = packageProfileList
                })
            Catch ex As Exception
                cp.Site.ErrorReport(ex)
            End Try
            '
            Return result
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' Constructor
        ''' </summary>
        ''' <remarks></remarks>
        Public Sub New(cp As CPBaseClass)
            Me.cp = cp
            '
            ' -- prepopulate request (lazy load later)
            allowPlace = cp.Doc.GetBoolean("AllowSelectResource")
            '
            ' -- topFolder should be in this format toptier\tier2\tier2
            ' -- all lowercase, no leading or trailing slashes, backslashs, remove 'root\'
            topFolderPath = cp.Doc.GetText("RootFolderName")
            topFolderPath = Trim(topFolderPath)
            topFolderPath = LCase(topFolderPath)
            topFolderPath = Replace(topFolderPath, "/", "\")
            If Left(topFolderPath, 4) = "root" Then
                topFolderPath = Mid(topFolderPath, 5)
            End If
            If Left(topFolderPath, 1) = "\" Then
                topFolderPath = Mid(topFolderPath, 2)
            End If
            If Right(topFolderPath, 1) = "\" Then
                topFolderPath = Mid(topFolderPath, 1, Len(topFolderPath) - 1)
            End If
            '
            AllowGroupAdd = cp.Doc.GetBoolean("AllowGroupAdd")
        End Sub
        '
        Public Shared Function serializeObject(ByVal CP As CPBaseClass, ByVal dataObject As Object) As String
            Dim result As String = ""
            Try
                Dim json_serializer As New System.Web.Script.Serialization.JavaScriptSerializer
                result = json_serializer.Serialize(dataObject)
            Catch ex As Exception
                CP.Site.ErrorReport(ex)
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' list of events and their stopwatch times
        ''' </summary>
        Public Class packageProfileClass
            Public name As String
            Public time As Integer
        End Class
        '
        '====================================================================================================
        ''' <summary>
        ''' remote method top level data structure
        ''' </summary>
        <Serializable()>
        Public Class packageClass
            Public success As Boolean = False
            Public errorList As New List(Of packageErrorClass)
            Public nodeList As New List(Of packageNodeClass)
            Public profileList As List(Of packageProfileClass)
        End Class
        '
        '====================================================================================================
        ''' <summary>
        ''' data store for jsonPackage
        ''' </summary>
        <Serializable()>
        Public Class packageNodeClass
            Public dataFor As String = ""
            Public data As Object ' IEnumerable(Of Object)
        End Class
        '
        '====================================================================================================
        ''' <summary>
        ''' error list for jsonPackage
        ''' </summary>
        <Serializable()>
        Public Class packageErrorClass
            Public number As Integer = 0
            Public description As String = ""
        End Class
        '
#Region " IDisposable Support "
        Protected disposed As Boolean = False
        '
        '==========================================================================================
        ''' <summary>
        ''' dispose
        ''' </summary>
        ''' <param name="disposing"></param>
        ''' <remarks></remarks>
        Protected Overridable Overloads Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposed Then
                If disposing Then
                    '
                    ' ----- call .dispose for managed objects
                    '

                End If
                '
                ' Add code here to release the unmanaged resource.
                '
            End If
            Me.disposed = True
        End Sub
        ' Do not change or add Overridable to these methods.
        ' Put cleanup code in Dispose(ByVal disposing As Boolean).
        Public Overloads Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
        Protected Overrides Sub Finalize()
            Dispose(False)
            MyBase.Finalize()
        End Sub
#End Region
    End Class
End Namespace

File: source/ResourceLibrary/Controllers/applicationController.cs

using System;
using System.Collections.Generic;
using System.Text.Json;
using Contensive.BaseClasses;

namespace Contensive.Addons.ResourceLibrary.Controllers {
    public class applicationController : IDisposable {
        private CPBaseClass cp;
        public bool allowPlace { get; }
        public string topFolderPath { get; private set; }
        public bool AllowGroupAdd { get; }
        public List<packageErrorClass> packageErrorList { get; set; } = new List<packageErrorClass>();
        public List<packageNodeClass> packageNodeList { get; set; } = new List<packageNodeClass>();
        public List<packageProfileClass> packageProfileList { get; set; } = new List<packageProfileClass>();

        public string getSerializedPackage() {
            string result = "";
            try {
                result = serializeObject(cp, new packageClass {
                    success = packageErrorList.Count.Equals(0),
                    nodeList = packageNodeList,
                    errorList = packageErrorList,
                    profileList = packageProfileList
                });
            } catch (Exception ex) {
                cp.Site.ErrorReport(ex);
            }
            return result;
        }

        public applicationController(CPBaseClass cp) {
            this.cp = cp;
            allowPlace = cp.Doc.GetBoolean("AllowSelectResource");
            topFolderPath = cp.Doc.GetText("RootFolderName");
            topFolderPath = topFolderPath.Trim();
            topFolderPath = topFolderPath.ToLower();
            topFolderPath = topFolderPath.Replace("/", "\\");
            if (topFolderPath.Length >= 4 && topFolderPath.Substring(0, 4) == "root") {
                topFolderPath = topFolderPath.Substring(4);
            }
            if (topFolderPath.Length >= 1 && topFolderPath.Substring(0, 1) == "\\") {
                topFolderPath = topFolderPath.Substring(1);
            }
            if (topFolderPath.Length >= 1 && topFolderPath.Substring(topFolderPath.Length - 1) == "\\") {
                topFolderPath = topFolderPath.Substring(0, topFolderPath.Length - 1);
            }
            AllowGroupAdd = cp.Doc.GetBoolean("AllowGroupAdd");
        }

        public static string serializeObject(CPBaseClass CP, object dataObject) {
            string result = "";
            try {
                result = JsonSerializer.Serialize(dataObject);
            } catch (Exception ex) {
                CP.Site.ErrorReport(ex);
            }
            return result;
        }

        public class packageProfileClass {
            public string name;
            public int time;
        }

        [Serializable]
        public class packageClass {
            public bool success = false;
            public List<packageErrorClass> errorList = new List<packageErrorClass>();
            public List<packageNodeClass> nodeList = new List<packageNodeClass>();
            public List<packageProfileClass> profileList;
        }

        [Serializable]
        public class packageNodeClass {
            public string dataFor = "";
            public object data;
        }

        [Serializable]
        public class packageErrorClass {
            public int number = 0;
            public string description = "";
        }

        #region  IDisposable Support
        protected bool disposed = false;
        protected virtual void Dispose(bool disposing) {
            if (!this.disposed) {
                if (disposing) {
                }
            }
            this.disposed = true;
        }
        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~applicationController() {
            Dispose(false);
        }
        #endregion
    }
}
